runtime/doc/lua.txt

```diff
@@ -717,6 +717,16 @@ vim.NIL								    *vim.NIL*
 	is equivalent to a missing value: `{"foo", nil}` is the same as 
 	`{"foo"}`
 
+vim.empty_dict()					    *vim.empty_dict()*
```

Comment:
Why not in a docstring ? ;_;

Reply:
I just simply wrote the documentation close to the other functions it belongs with (rpcrequest et al)? Or is there a way to write a docstring and still have it show up at a certain place in `|lua-builtin|` and not with unrelated `|lua-vim|` stuff? 

Or should `lua-builtin` and `lua-vim` be merged perhaps, the distinction might be quite arbitrary to the end user (both c and lua code inside the binary are "builtin").

Reply:
> Or should `lua-builtin` and `lua-vim` be merged perhaps, the distinction might be quite arbitrary to the end user (both c and lua code inside the binary are "builtin").

The `lua-builtin` section is just a hack/workaround for functions that are impl'd in C. (Maybe we could doc those in a vim.lua stub? Then they could be generated and we can eliminate lua-builtin section.)

> Or is there a way to write a docstring and still have it show up at a certain place in |lua-builtin| and not with unrelated |lua-vim| stuff?

There's no way to group stuff, currently the docs are generated by alphabetical order.

If grouping stuff is wanted I would suggest something analogous to `:help reference_toc` which serves as a grouped view of various disconnected things.

Reply:
> There's no way to group stuff, currently the docs are generated by alphabetical order. If grouping is wanted I would suggest something analogous to :help reference_toc which serves as a grouped view of various disconnected things.

I certainly did not speak of any "want" to create groupings, just the reality of the structure that already exists _today_. If `|lua-builtin|` cannot be defined by a docstring, then that answers your question why this couldn't be a docstring right now.

> Maybe we could stub those in vim.lua (then reassign the stubs in executor.c)? Then they could be generated and we can eliminate lua-builtin section.

This would be an improvement , or I actually though of docs strings on the C impl functions inside executor.c directly, not sure what is simplest to implement.

Reply:
>  If `|lua-builtin|` cannot be defined by a docstring, then that answers your question why this couldn't be a docstring right now.

Well, only if one cares about grouping it there. Otherwise it could be a docstring here: https://github.com/neovim/neovim/blob/0c436559dfba9e0bb7e6ce5a835ed8eee540205f/src/nvim/lua/vim.lua#L246

> or I actually though of docs strings on the C impl functions inside executor.c directly, not sure what is simplest to implement.

sure, that's fine too. We can add some "routing" to `gen_vimdoc.py` fairly easily.

Reply:
> Well, only if one cares about grouping it there.

If one does not care there is no reason to move it either way.

> lua-builtin is just a place for stuff that doesn't exist explicitly in vim.lua. But in this case, empty_dict() does exist in vim.lua, so it could be doc'd there.

Technically true but missing the point. `vim.empty_dict()` is _exactly_ as much a builtin concept as `vim.NIL`. This concept just happens to be named with a two-lined helper function in `vim.lua` that creates an instance (unlike `vim.NIL` which is a singleton). Even instances created by builtin functions will be represented by the string `"vim.empty_dict()"`. Segregating `vim.empty_dict()` away from the "real" builtin stuff just because it has a two-line "shell" apart from all the internal plumbing (which is the stuff that really makes it work as a "type") would be quite confusing. While merging would probably be the best, as long as the distinction still exists it does matter, and we shouldn't place stuff in the less optimal category just because it is technically possible to do so.

Reply:
My point was only to use docstrings as much as possible, because

1. It generates better formatted docs with less maintenance
2. It is included in things like lua.mpack

I know it's ugly that some things are separate, but I'm hoping that is a temporary situation and would like to avoid accumulating more non-docstring cases. That's all.

Reply:
> My point was only to use docstrings as much as possible, because
    It generates better formatted docs with less maintenance
    It is included in things like lua.mpack

I agree with all of this. But we can only achieve this by _systematic_ improvements so that the separation can be eliminated, rather than changing the trade-off in individual border cases, which certainly will not stop "accumulating more non-docstring cases" (unless we have a ban on adding bulitin functionality). An actual solution would be something that enables _all_ builtins to be placed in `lua.mpack`, only having `vim.empty_dict()` in there but not its related functions/objects such as `vim.NIL` does not solve anything.